File: research/portotipos/simple_iq_plot.py
import os
import numpy as np
import matplotlib.pyplot as plt

# 1. Identificación del Proyecto: UIC Radiotelescopio (Plataforma DSP para Radioastronomía)
# Fichero de datos IQ pre-grabado por defecto en formato int16 (I: 16-bit, Q: 16-bit)
FILE_PATH = os.path.join("data", "test_signal.iq")
SAMPLE_RATE = 10_000_000  # 10 MSps (ejemplo)
WINDOW_SIZE = 50          # Ancho del Moving Average

# 2. Leer muestras del archivo IQ
# Cada muestra compleja consta de I (int16) y Q (int16) -> 4 bytes por muestra
num_samples = 2000
bytes_to_read = num_samples * 2 * 2  # 2 canales * 2 bytes/muestra

if not os.path.exists(FILE_PATH):
    raise FileNotFoundError(f"No se encontró el archivo de datos en {FILE_PATH}")

with open(FILE_PATH, "rb") as f:
    raw_data = f.read(bytes_to_read)

# Decodificar int16 y normalizar a rango [-1.0, 1.0]
samples = np.frombuffer(raw_data, dtype=np.int16).astype(np.float32) / 32768.0
iq_raw = samples[0::2] + 1j * samples[1::2]
t = np.arange(len(iq_raw)) / SAMPLE_RATE

# 3. Aplicar Filtro Moving Average (Promedio Móvil)
# Se aplica np.convolve de forma directa sobre la señal compleja
window = np.ones(WINDOW_SIZE) / WINDOW_SIZE
iq_filtered = np.convolve(iq_raw, window, mode='same')

# 4. Graficar Señales
fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 6), sharex=True)

# Gráfica Superior: Señal IQ Cruda
ax1.plot(t * 1e6, iq_raw.real, label="Componente I (Real)", color="cyan", alpha=0.7)
ax1.plot(t * 1e6, iq_raw.imag, label="Componente Q (Imaginario)", color="magenta", alpha=0.7)
ax1.set_title("Señal IQ Original (Sin Filtrar)")
ax1.set_ylabel("Amplitud (V)")
ax1.legend(loc="upper right")
ax1.grid(True, linestyle="--", alpha=0.5)

# Gráfica Inferior: Señal Filtrada (Moving Average)
ax2.plot(t * 1e6, iq_filtered.real, label="Componente I Filtrado", color="green", alpha=0.9)
ax2.plot(t * 1e6, iq_filtered.imag, label="Componente Q Filtrado", color="orange", alpha=0.9)
ax2.set_title(f"Señal Filtrada - Filtro Moving Average (Ventana = {WINDOW_SIZE})")
ax2.set_xlabel("Tiempo (μs)")
ax2.set_ylabel("Amplitud (V)")
ax2.legend(loc="upper right")
ax2.grid(True, linestyle="--", alpha=0.5)

plt.tight_layout()
plt.show()

File: research/portotipos/simple_iq_plot.m
clear all

filePath = fullfile('data', 'test_signal.iq');
sampleRate = 2e6;     
windowSize = 50;       
numSamples = 2000;      

if ~exist(filePath, 'file')
    error('No se encontró el archivo de datos en %s', filePath);
end

% 2. Leer muestras del archivo IQ
fileID = fopen(filePath, 'r');
rawSamples = fread(fileID, numSamples * 2, 'int16');
fclose(fileID);

% Separar canales I y Q, y normalizar
samplesNormalized = double(rawSamples) / 32768.0;
I = samplesNormalized(1:2:end);
Q = samplesNormalized(2:2:end);
iqRaw = I + 1i*Q;

N = length(iqRaw);
samplesAxis = 0:N-1; % Eje de muestras

% 3. Aplicar Filtro Moving Average (Promedio Móvil)
iqFiltered = movmean(iqRaw, windowSize);

% 4. Calcular Espectros de Potencia (FFT)
f = linspace(-sampleRate/2, sampleRate/2, N) / 1e6; % Frecuencia en MHz

% FFT con ventana de Hann para reducir fugas espectrales
windowHann = hann(N);
fftRaw = fftshift(fft(iqRaw .* windowHann));
fftFiltered = fftshift(fft(iqFiltered .* windowHann));

% Espectros en dBFS
psdRaw = 10 * log10((abs(fftRaw).^2) / (N * sum(windowHann.^2)) + 1e-12);
psdFiltered = 10 * log10((abs(fftFiltered).^2) / (N * sum(windowHann.^2)) + 1e-12);

% 5. Graficar (Layout 2x2)
figure('Color', [1 1 1], 'Name', 'Análisis Temporal y Espectral - Plataforma DSP');

% --- Fila 1: Señal Original ---
% Temporal Original (Eje X en Muestras)
subplot(2, 2, 1);
plot(samplesAxis, real(iqRaw), 'Color', [0 0.7 0.9], 'LineWidth', 1.0); hold on;
plot(samplesAxis, imag(iqRaw), 'Color', [0.9 0 0.6], 'LineWidth', 1.0);
title('Señal IQ Original (Tiempo)');
xlabel('Número de Muestra (n)');
ylabel('Amplitud (V)');
legend('I (Real)', 'Q (Imaginario)');
grid on;

% Espectro Original
subplot(2, 2, 2);
plot(f, psdRaw, 'Color', [0 0.5 0.8], 'LineWidth', 1.0);
title('Espectro de Frecuencia Original');
xlabel('Frecuencia (MHz)');
ylabel('Potencia (dBFS)');
grid on;

% --- Fila 2: Señal Filtrada (Moving Average) ---
% Temporal Filtrado (Eje X en Muestras)
subplot(2, 2, 3);
plot(samplesAxis, real(iqFiltered), 'Color', [0 0.6 0], 'LineWidth', 1.2); hold on;
plot(samplesAxis, imag(iqFiltered), 'Color', [0.9 0.5 0], 'LineWidth', 1.2);
title(sprintf('Señal IQ Filtrada (MA: %d)', windowSize));
xlabel('Número de Muestra (n)');
ylabel('Amplitud (V)');
legend('I Filtrado', 'Q Filtrado');
grid on;

% Espectro Filtrado
subplot(2, 2, 4);
plot(f, psdFiltered, 'Color', [0.2 0.7 0.2], 'LineWidth', 1.2);
title('Espectro de Frecuencia Filtrado');
xlabel('Frecuencia (MHz)');
ylabel('Potencia (dBFS)');
grid on;
